Docsp 8721 mongoose vs driver pg (#717)
mongodb/docs-node: source/driver-vs-mongoose.txt

to your collection. However, the biggest difference between the two are
the way they each set and manage schema validation.

The {+driver-long+} applies the schema validation at the database
level. In comparison, Mongoose applies schema validation at the application level. In the
{+driver-short+} schema validation works on individual collections by
using either `JSON Schema Validation <https://www.mongodb.com/docs/manual/core/schema-validation/specify-json-schema/>`__
or `Schema Validation with Query Operators <https://www.mongodb.com/docs/manual/core/schema-validation/specify-query-expression-rules/>`__.
Validation with query operators is useful for dynamic validation rules
calculated at runtime. By contrast, JSON validation is useful for, among other
reasons, being an open standard which makes it easy to connect to
third-party libraries.

Mongoose applies the schema validation in a twofold procedure; firstly by telling your
application what data types and fields are allowed to be in a particular
collection, then it confirms the data matches the set criteria. Mongoose's
schema validation works on the application level, meaning MongoDB

> doesn't know it exists.

Reviewer: This reads a little weird. Maybe something like:

_"Unlike the {+driver-long+}, Mongoose's schema validation occurs client-side. Schema validation is applied by first telling your application what data types and fields are allowed to be in a particular collection, then by confirming the data matches the set criteria."_

.. tip::
   
   Though Mongoose can be more efficient to create schemas,
   the {+driver-long+} is generally twice as fast at reading and writing
   operations. This means for larger, more performance-intensive
   projects the {+driver-short+} is likely the better option.

Similarities
------------

As Mongoose is built upon the {+driver-long+}, they share many
capabilities. Even if you use Mongoose, you always have the option to
use any native function or command built into the {+driver-short+}

> alongside Mongoose. Mongoose also offers its own methods that are quite

Reviewer: >  Mongoose also offers its own methods

 Mongoose also offers its own APIs

> alongside Mongoose. Mongoose also offers its own methods that are quite

Reviewer: >  Mongoose also offers its own methods

 Mongoose also offers its own APIs

similar to the {+driver-short+}, such as ``updateOne()`` and
``replaceOne()``, though these methods work differently under the hood.

.. tip::
   
   There are many other Mongoose specific methods not discussed. For
   more information on how to use Mongoose, check out the official `Mongoose
   documentation <https://mongoosejs.com/docs/>`__.

Differences
-----------
Though MongoDB was designed with a flexible data model in mind, Mongoose
forces a semi-rigid ``schema`` from the beginning. With Mongoose, developers
must define a ``schema`` and ``model``. Moreover, the handling of read
and write operations differs between the two, resulting in the

> {+driver-short+} generally running twice as fast as Mongoose. 

Reviewer: > "Moreover, the handling of read and write operations differs between the two, resulting in the {+driver-short+} generally running twice as fast as Mongoose."

I would cut that entirely as it's not accurate

.. _node-driver-vs-mongoose:

Comparing the Node.js Driver and Mongoose
=========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Overview
--------

On this page, you can learn about the comparison between
the official MongoDB Node.js driver and the third-party library called
Mongoose.

The MongoDB Node.js driver is the officially developed, supported, and maintained
MongoDB driver for Node.js. 

Moongoose is a popular community-developed Object Data Modeling (ODM)
library that is built on top of the {+driver-short+}. Though you don't
need to use an ODM tool to work with MongoDB, some developers prefer

> them as it is similar to working with a relational database. 

Reviewer: FYI that ODM is typically used for "Object-Document Mapper" (the nosql variant of ORM - Object Relational Mapper)
ex:
- https://www.mongodb.com/docs/drivers/ruby/
- https://www.doctrine-project.org/projects/doctrine-mongodb-odm/en/latest/
- https://www.udemy.com/course/mongoose/

I can see the other usage at https://www.mongodb.com/tools/mongoose though so I'm not sure where that reference originally came from

> them as it is similar to working with a relational database. 

Reviewer: > "... as it is similar to working with a relational database."
"... as it is similar to using an ORM which they've likely used previously when working with a relational database."

capabilities. Even if you use Mongoose, you always have the option to
use any native function or command built into the {+driver-short+}
alongside Mongoose. Mongoose also offers its own methods that are quite
similar to the {+driver-short+}, such as ``updateOne()`` and
``replaceOne()``, though these methods work differently under the hood.

Differences
-----------
Though MongoDB was designed with a flexible data model in mind, Mongoose
forces a semi-rigid ``schema`` from the beginning. With Mongoose, developers
must define a ``schema`` and ``model``. Moreover, the handling of read
and write operations differs between the two, resulting in the
{+driver-short+} generally running twice as fast as Mongoose. 

.. tip::
   
   There are many other Mongoose specific methods not discussed. For
   more information on how to use Mongoose, check out the official `Mongoose
   documentation <https://mongoosejs.com/docs/>`__.

Conclusion
----------

Both Mongoose and the {+driver-long+} offer comparable features
and capabilities. While Mongoose helps ease MongoDB development for
newcomers, it employs various automated processes behind the scenes that

> might lead to unforeseen consequences down the line.

Reviewer: >  While Mongoose helps ease MongoDB development for
newcomers, it employs various automated processes behind the scenes that
might lead to unforeseen consequences down the line.

This is framing mongoose as a risk that developers should avoid. Mongoose is part of the [Verified Solutions Program](https://www.mongodb.com/blog/post/mongodb-announces-new-verified-solutions-program) which implies customers with commercial support can officially use mongoose and receive support for their applications.

The conclusion should be used to promote the value of mongoose as a product while calling out that it adds a layer of abstraction above the driver - not that it will result in a negative experience.

Reviewer: It might be worth adding a link to the [Mongoose Learning Byte](https://learn.mongodb.com/learn/course/mongoose-odm-with-mongodb/learning-byte/learn) as well

.. _node-driver-vs-mongoose:

Comparing the Node.js Driver and Mongoose
=========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Overview
--------

In this page, you can identify the differences and similarities between
the official MongoDB Node.js driver and the third-party library called
Mongoose.

> The MongoDB Node.js driver is the officially developed, supported, and maintained

Author: I personally prefer to highlight the fact that the native driver is _the_ MongoDB supported driver, but if this is a hard-stop for you I can change!

MongoDB driver for Node.js. The Node.js driver is a library of functions
that you can use to connect to and communicate with MongoDB.

Moongoose is a popular community-developed Object Data Modeling (ODM)
library that is built on top of the {+driver-short+}. While you don't
need to use an ODM tool to work with MongoDB, some developers prefer them. 

This page contains a high-level comparison of most of the ways the two differ.

Schema Validation
-----------------

Both the {+driver-long+} and Mongoose lets you apply schema validation
to your collection. However, the biggest difference between the two are

> the way they each handle and manage schema validation.

Author: I was trying to highlight the fact that they differ in both the way schemas are set and the way are handled but I should probably state it exactly like that!

While the {+driver-long+} applies the schema validation at the database level, Mongoose
applies schema validation at the application level. In the
{+driver-short+} schema validation works on individual collections by
using either `JSON Schema Validation
<https://www.mongodb.com/docs/manual/core/schema-validation/specify-json-schema/>`__
or `Schema Validation with Query Operators
<https://www.mongodb.com/docs/manual/core/schema-validation/specify-query-expression-rules/>`__.
Validation with query operators is useful for dynamic validation rules

> calculated at runtime, while JSON validation is useful for, among other

Reviewer 2: ```suggestion
calculated at runtime. By contrast, JSON validation is useful for, among other
```

reasons, being an open standard which makes it easy to connect to
third-party libraries.

Mongoose applies the schema validation in a twofold procedure; firstly by telling your
application what data types and fields are allowed to be in a particular
collection, then it confirms the data matches the set criteria. Mongoose's
schema validation works on the application level, meaning MongoDB
doesn't know it exists.

.. tip::
   
   While Mongoose can be more efficient to create schemas,
   the {+driver-long+} is generally twice as fast at reading and writing
   operations. This means for larger, more performance-intensive
   projects the {+driver-short+} is likely the better option.

Similarities
------------

> Since Mongoose is built upon the {+driver-long+}, they share many

Reviewer 2: ```suggestion
As Mongoose is built upon the {+driver-long+}, they share many
```

   :class: singlecol

Overview
--------

On this page, you can learn about the comparison between
the official MongoDB Node.js driver and the third-party library called
Mongoose.

The MongoDB Node.js official driver is a library of functions used to
connect to and communicate with MongoDB. 

Moongoose is a popular community-developed Object-Document Mapper (ODM)
library that is built on top of the {+driver-short+}. Though you don't
need to use an ODM tool to work with MongoDB, some developers prefer
them as it is similar to using an Object-Relational Mapper, or ORM,
which they've likely used when working with a relational database.

The official MongoDB native driver provides direct low-level access to
MongoDB, allows for more flexibility and control but requires more
manual effort for data modeling and validation. On the other hand,
Mongoose abstracts away many complexities, providing data modeling,
schema validation, middleware support, and other convenient features,
making it easier to work with MongoDB in Node.js applications. The
choice between the two depends on your specific project requirements and
how much control and flexibility you need in database interactions.

> 

Reviewer 3: S: Before going into the sections, this page should include a table of contents describing the sections at a high level
Similar to this [one](https://github.com/mongodb/docs-rust/blob/f2d475c8238129de65955fe48289e668852e0b97/source/fundamentals/crud/compound-operations.txt#L33-L49)

the database. With Mongoose, you can define data schemas and models,
allowing you to work with MongoDB in a more structured and intuitive
manner. 

Differences in Schema Management
--------------------------------

Both the {+driver-long+} and Mongoose let you apply schema validation
to your collection. However, the biggest difference between the two are
the way they each set and manage schema validation.

Mongoose allows you to define data schemas using its schema notation.
These schemas define the structure of the documents to be stored in
MongoDB, along with data types, default values, and validation rules.
Mongoose provides built-in support for data validation based on the
defined schema rules. It ensures that the data being saved to the
database adheres to the specified schema, preventing invalid data from
being stored. 

The {+driver-long+} applies the schema validation at the database
level. In comparison, Mongoose applies schema validation at the application level. In the
{+driver-short+} schema validation works on individual collections by
using either `JSON Schema Validation <https://www.mongodb.com/docs/manual/core/schema-validation/specify-json-schema/>`__

> or `Schema Validation with Query Operators <https://www.mongodb.com/docs/manual/core/schema-validation/specify-query-expression-rules/>`__.

Reviewer 3: S: use the :manual: role whenever you link manual pages. Alternatively, move these links to the end of the paragraph as follows:

...

To learn more about XXX, see :manual:`xyz <>` in the Server manual.

Validation with query operators is useful for dynamic validation rules
calculated at runtime. By contrast, JSON validation is useful for, among other

> reasons, being an open standard which makes it easy to connect to

Reviewer 3: Q: what is meant by "open standard"?

Author: By saying open standard, I mean to say JSON is a standard that is openly accessible and usable by anyone that uses an open license that provides for extensibility.

third-party libraries.

Similarities
------------

> As Mongoose is built upon the {+driver-long+}, they share many
> capabilities. Even if you use Mongoose, you always have the option to

Reviewer 3: ```suggestion
Because Mongoose is built on the {+driver-short+}, they share many
functionalities. Even if you use Mongoose, you always have the option to
```

> capabilities. Even if you use Mongoose, you always have the option to
> use any native function or command built into the {+driver-short+}
> alongside Mongoose. Mongoose also offers its own methods that are quite
> similar to the {+driver-short+}, such as ``updateOne()`` and
> ``replaceOne()``.

Reviewer 3: ```suggestion
If you use Mongoose, you can also integrate any APIs from the {+driver-short+}.
Mongoose also offers its own APIs that are similar to those provided by the driver, including ways to update and replace documents.
```

Reviewer: @rustagir in this case since API similarities are being called out wouldn't it be better to keep the methods in there (`updateOne()`) as opposed to generalizing it to "ways to update ..."?

.. tip::
   

>    There are many other Mongoose specific methods and APIs not discussed. For
>    more information on how to use Mongoose, check out the official `Mongoose
>    documentation <https://mongoosejs.com/docs/>`__.

Reviewer 3: ```suggestion
   This guide does not describe all of the Mongoose APIs. For
   detailed information these APIs, see the `official Mongoose
   documentation <https://mongoosejs.com/docs/>`__.
```

> Additional Functionality Provided by Mongoose
> ---------------------------------------------

Reviewer 3: ```suggestion
Other Mongoose Features
-----------------------
```

> On top of schema management, there is a multitude of additional
> functionality provided by Mongoose. In this section, we will go over a
> few of the additional functionalities.

Reviewer 3: ```suggestion
This section describes additional features available in Mongoose beyond schema validation.
```

Data Modeling
~~~~~~~~~~~~~

> Mongoose introduces the concept of models, which are JavaScript classes
> defined based on schemas. Models act as an interface to interact
> with MongoDB collections, abstracting away the complexities of database
> operations.

Reviewer 3: ```suggestion
Mongoose uses *models*, JavaScript classes that you define from schemas. You can use models to interact
with documents in your MongoDB collections, abstracting away the complexities of database
operations.
```

> operations.

Reviewer 3: Q: how is this achieved? If you can't explain, it is best to leave this out

Middleware Hooks
~~~~~~~~~~~~~~~~

> Mongoose supports middleware hooks, both pre and post-hooks, that allow you

Reviewer 3: Q: what are pre- and post-hooks? If it is not relevant, leave out

Author: pre- and post-hooks is another name for middleware, but it makes sense to leave it out.

> to add custom logic before or after certain database operations like

Reviewer 3: Q: what does the custom logic do? can you provide an example?

Author: there are tons of uses and examples of the custom logic, so I was weary to add an example to avoid making the user think it has a narrow use case, but I've included one now.

> to add custom logic before or after certain database operations like
> save, update, and remove. This is useful for implementing additional

Reviewer 3: ```suggestion
to add custom logic before or after CRUD operations, including update and delete operations. This is useful for implementing additional
```

actions or validations. Mongoose supports adding custom plugins and

> middleware functions to extend its functionality further. 

Reviewer 3: Q: similar to above question -- what actions or validations? What are these custom plugins/functions for?

Author: Similar answer to the above question. There are a multitude of actions or validations available for use with middleware and the custom plugins are outside the scope of Mongoose's official offerings so I didn't think to leave an example. I think it would be better to link to the Mongoose documentation on middleware so the user can see all the uses if they want to learn more but am open to other opinions.

Virtuals
~~~~~~~~

> In Mongoose, a virtual is a property that is not stored in MongoDB.

Reviewer 3: ```suggestion
Mongoose supports *virtuals*, properties that are not stored in MongoDB.
```

> Virtuals are typically used for computed properties on
> documents. Virtuals are additional fields for a given model; their values
> can be set manually or automatically with defined functionality. They
> only exist logically and are not written to the document, collection, or
> database.

Reviewer 3: ```suggestion
You can create virtuals to store computed properties from other document fields.
Virtuals are additional fields for a given model, and you can set their values
manually or automatically through defined functions.  They
only exist within the logic of the application and are not written to the document, collection, or
database.
```

> Conclusion

Reviewer 3: S: this section seems more like a "Summary" than a conclusion -- as it is not making an argument

----------

> Both Mongoose and the {+driver-long+} offer comparable features
> and capabilities. Mongoose helps ease MongoDB development for

Reviewer 3: ```suggestion
Mongoose and the {+driver-short+} provide comparable features
and capabilities. Mongoose helps ease MongoDB development for
```

> and capabilities. Mongoose helps ease MongoDB development for
> newcomers while the {+driver-long+} seamlessly integrates with your MongoDB
> database, providing an exceptional and adaptable development experience.
> By leveraging the driver, you enable your database to excel at its core
> functionality.

Reviewer 3: ```suggestion
and capabilities. Mongoose can ease the development of applications for
new users of MongoDB. Because the {+driver-short+} is actively maintained and improved by MongoDB, 
the driver seamlessly integrates with your MongoDB
database, providing a straightforward and adaptable development experience.
```

> .. tip::

Reviewer 3: S: move this content out of a tip into a section called "Next Steps" or something similar

   
   To learn how to start using the {+driver-long+}, visit the
   :ref:`Quick Start guide <node-quickstart>`. To learn how to get
   started with Mongoose, see the `Getting Started with MongoDB &
   Mongoose
   <https://www.mongodb.com/developer/languages/javascript/getting-started-with-mongodb-and-mongoose/>`__
   page and the `Mongoose ODM with MongoDB Learning Byte

>    <https://learn.mongodb.com/learn/course/mongoose-odm-with-mongodb/learning-byte/learn>`__.

Reviewer 3: S: use roles where possible, there is a :website: [role](https://github.com/mongodb/snooty-parser/blob/5dab215d105fa01976c6521c356323725e137a9d/snooty/rstspec-consolidated.toml#L1332-L1334)
